server/src/test/java/org/elasticsearch/cluster/coordination/Zen1IT.java

```diff
@@ -45,4 +66,198 @@ public void testZen1NodesJoiningZen2Cluster() {
+            }
+
+            internalCluster().restartNode(node, new RestartCallback() {
+                @Override
+                public Settings onNodeStopped(String restartingNode) {
+                    String viaNode = randomValueOtherThan(restartingNode, () -> randomFrom(nodes));
+                    final ClusterHealthRequestBuilder clusterHealthRequestBuilder = client(viaNode).admin().cluster().prepareHealth()
+                        .setWaitForEvents(Priority.LANGUID)
+                        .setWaitForNodes(Integer.toString(zen1NodeCount + zen2NodeCount - 1))
+                        .setTimeout(TimeValue.timeValueSeconds(30));
+                    ClusterHealthResponse clusterHealthResponse = clusterHealthRequestBuilder.get();
+                    assertFalse(restartingNode, clusterHealthResponse.isTimedOut());
+                    return Settings.EMPTY;
+                }
+            });
+            ensureStableCluster(zen1NodeCount + zen2NodeCount);
+            ensureGreen("test");
+
+            if (requiresVotingConfigExclusions) {
+                final ClearVotingConfigExclusionsRequest clearVotingTombstonesRequest = new ClearVotingConfigExclusionsRequest();
+                clearVotingTombstonesRequest.setWaitForRemoval(false);
+                client().execute(ClearVotingConfigExclusionsAction.INSTANCE, clearVotingTombstonesRequest).get();
+            }
+        }
+    }
+
+    private void testRollingMigrationFromZen1ToZen2(final int nodeCount) throws Exception {
+        final List<String> zen1Nodes = internalCluster().startNodes(nodeCount, ZEN1_SETTINGS);
+
+        createIndex("test",
+            Settings.builder()
+                .put(UnassignedInfo.INDEX_DELAYED_NODE_LEFT_TIMEOUT_SETTING.getKey(), TimeValue.ZERO) // assign shards
+                .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, nodeCount) // causes rebalancing
+                .put(IndexMetaData.SETTING_NUMBER_OF_REPLICAS, 1)
+                .build());
+        ensureGreen("test");
+
+        for (final String zen1Node : zen1Nodes) {
+            logger.info("--> shutting down {}", zen1Node);
+            internalCluster().stopRandomNode(s -> NODE_NAME_SETTING.get(s).equals(zen1Node));
+
+            ensureStableCluster(nodeCount - 1);
+            if (nodeCount > 2) {
+                ensureGreen("test");
+            } else {
+                ensureYellow("test");
+            }
+
+            logger.info("--> starting replacement for {}", zen1Node);
+            final String newNode = internalCluster().startNode(ZEN2_SETTINGS);
+            ensureStableCluster(nodeCount);
+            ensureGreen("test");
+            logger.info("--> successfully replaced {} with {}", zen1Node, newNode);
+        }
+
+        assertThat(internalCluster().size(), equalTo(nodeCount));
+    }
+
+    public void testMigratingFromZen1ToZen2ClusterWithTwoNodes() throws Exception {
+        testRollingMigrationFromZen1ToZen2(2);
```

Comment:
as these test are somewhat slow, is it sufficient to have one of these and randomize the count?